haystack/nodes/label_generator/pseudo_label_generator.py

```diff
@@ -0,0 +1,211 @@
+import sys
+import random
+from typing import List, Dict, Union, Tuple
+
+from sentence_transformers import CrossEncoder
+
+from haystack.nodes import BaseRetriever, QuestionGenerator
+from haystack.schema import Document
+from haystack.nodes.base import BaseComponent
+
+
+class PseudoLabelGenerator(BaseComponent):
+    """
+    The PseudoLabelGenerator is a component that creates Generative Pseudo Labeling (GPL) training data for training
+    of dense retrievers.
+
+    GPL is an unsupervised domain adaptation method for training of dense retrievers. It is based on question generation
+    and pseudo labeling with powerful cross-encoders. To train a domain-adapted model, it needs access to unlabeled
+    target corpus usually via DocumentStore and a retriever to mine for negatives.
+
+    For more details see https://github.com/UKPLab/gpl
+
+    For example:
+
+    ```python
+    |   document_store = DocumentStore(...)
+    |   retriever = Retriever(...)
+    |   qg = QuestionGenerator(model_name_or_path="doc2query/msmarco-t5-base-v1")
+    |   plg = PseudoLabelGenerator(qg, retriever)
+    |   train_examples = []
+    |   for idx, doc in enumerate(document_store):
+    |   output_samples = plg.run(documents=[doc])
+    |   for item in output_samples:
+    |       train_examples.append(item)
+    |
+    |   # Users can also provide a list of questions/document dictionary pairs to generate pseudo labels for
+    |
+    |   question_doc_pairs: List[Dict[str, str]] = [{"question": "question text ...", "document": "document text ..."}]
+    |   plg = PseudoLabelGenerator(question_doc_pairs, retriever)
+    |   train_examples = []
+    |   for idx, doc in enumerate(document_store):
+    |   output_samples = plg.run(documents=[doc])
+    |   for item in output_samples:
+    |       train_examples.append(item)
+    ```
+    """
+
+    def __init__(
+        self,
+        question_producer: Union[QuestionGenerator, List[Dict[str, str]]],
+        retriever: BaseRetriever,
+        cross_encoder_model_name_or_path: str = "cross-encoder/ms-marco-MiniLM-L-6-v2",
+        max_questions_per_document: int = 3,
+        top_k: int = 50,
```

Comment:
Maybe we can add an optional `batch_size` parameter here that is used in `generate_questions`,  `mine_negatives` and `generate_margin_scores`.